### ISSUE TITLE
Integrate OpenCV's image detection into zm_monitor.cpp as an alternate option

### ISSUE THREAD

**Issue by pliablepixels** [ZoneMinder/zoneminder#2138]
OpenCV offers some "reasonable" methods to detect people. They are not as accurate as CNNs and use HOGs but are of orders of magnitude faster. The code is straight forward, for example, [this](https://github.com/opencv/opencv/blob/master/samples/cpp/peopledetect.cpp). OpenCV code is very portable, and in C++, and therefore integration into ZM is not hard. (I don't feel like investigating - someone who knows zm_monitor.cpp and how ZM handles image buffers should be able to do it much faster than me)

There are multiple profiles to choose from (different XML files)

**The request:** 
a) Integrate OpenCV with ZM 
b) Select "Person Detection" as a scheme as a detection algorithm as an alternate to "Pixels/Filter/Blob" and basically replicate the functionality provided in the link above.
c) Ability to specify a "frame interval" -> which will basically decide how often this logic will be used. In other words, even if I select the monitor FPS as say 5, I should be able to select "2" here, which will mean detection to be applied on the 1st, 3rd, 5th frames - for more speed control without affecting live fps
d) Draw a bounding box around the person detected in frame (Again, already part of the sample code)

**3rd party work**
a) [This company did some early work here](https://bitbucket.org/CVisionLab/zoneminder/wiki/changes) in 2011, if it helps. [Licensed under GPL](https://bitbucket.org/CVisionLab/zoneminder/src/93de44cc00fc4b58348a4ae41ca49c0c3b13dac3/COPYING?at=default&fileviewer=file-view-default).

**Future Enhancement**
Not a necessary condition for the bounty:
a) Ability to make item b) above in request section more flexible. Let me select any XML profile in OpenCV so I can tweak my model.

I'll be putting in a $100 bounty for this.

**Comment by pliablepixels** [comment]
bounty added

**Comment by connortechnology** [comment]
So I have merge the CVisionLabs work into my opencv branch.  However the plugin code is for an ancient version of opencv.... so I think I'll be updating that to version 3.

**Comment by pliablepixels** [comment]
do you have a branch I can track/test?

**Comment by connortechnology** [comment]
my opencv branch

**Comment by TheSwert** [comment]
Since version 3.4.2 of OpenCV, it looks like it now supports the YOLOv3 DNN detector as well. Below link has some C++ and python examples.
https://www.learnopencv.com/deep-learning-based-object-detection-using-yolov3-with-opencv-python-c/

**Comment by pliablepixels** [comment]
I'm going to revise some of the asks based on my experience [integrating both OpenCV approaches](https://github.com/pliablepixels/zmeventserver/tree/master/hook_example) in the ZM event notification server:

I originally said:

> b) Select "Person Detection" as a scheme as a detection algorithm as an alternate to "Pixels/Filter/Blob" and basically replicate the functionality provided in the link above.
> c) Ability to specify a "frame interval" -> which will basically decide how often this logic will be used. In other words, even if I select the monitor FPS as say 5, I should be able to select "2" here, which will mean detection to be applied on the 1st, 3rd, 5th frames - for more speed control without affecting live fps

I don't think this is practical.

Based on my experience,
a) Many (if not most) users of ZM are not running a GPU
b) OpenCV SVM HOG is very fast but not quite accurate. It keeps identifying my water heater as a "person"
c) OpenCV DNN with Yolov3 (80 labels) is quite accurate but without a GPU takes 10x the time as HOG (On an Intel Xeon 3.16GHz 4Core machine with no GPU HOG takes 0.25s and YoloV3 takes 2.4s)

To make this practically useful, I'd like to suggest that "Person detection" is implemented as a "post analysis" option to "motion" not as an alternate.

In other words, ZM first detect motion, using whichever means and _then_ runs person detection before an alarm triggers. This is the approach I am following, and even with a 2s analysis, its quite acceptable because it only kicks in _after_ ZM decides an alarm has occurred.

**Comment by 3lk0k0** [comment]
anyone know if these if going to do ?

**Comment by connortechnology** [comment]
I started on it... but there is lots more to do.

**Comment by jysaloma** [comment]
There is also some work done in this project for face recognition https://github.com/goruck/smart-zoneminder by goruck

**Comment by cleverinx** [comment]
Is this still in progress?

**Comment by connortechnology** [comment]
bountysource will take this money on July 1st.  Either contact bountysource to redirect your funds, or alternatively, I can close this out, collect the bounty, and I will hold the money until this issue is properly closed.

**Comment by pliablepixels** [comment]
I’d prefer you just collect for the money I added.

**Comment by Rayn0r** [comment]
@connortechnology 
Is implementing OpenCV into Zoneminder still on the list of things to do?

**Comment by connortechnology** [comment]
Yes

**Comment by connortechnology** [comment]
Yes it's still open. I'm not sure it's a good idea, as I think AI should be happening in it's own process.

**Comment by singhharsh1708** [comment]
@connortechnology is the bounty still open on this?

**Comment by connortechnology** [comment]
Closed to shutdown the annoying AI bots.  This issue has already been implemented.
